llama_index/indices/multi_modal/retriever.py

```diff
@@ -0,0 +1,117 @@
+            See reference for VectorStoreQueryMode for full list of supported modes.
+        filters (Optional[MetadataFilters]): metadata filters, defaults to None
+        alpha (float): weight for sparse/dense retrieval, only used for
+            hybrid query mode.
+        doc_ids (Optional[List[str]]): list of documents to constrain search.
+        vector_store_kwargs (dict): Additional vector store specific kwargs to pass
+            through to the vector store at query time.
+
+    """
+
+    def __init__(
+        self,
+        index: MultiModalVectorStoreIndex,
+        similarity_top_k: int = DEFAULT_SIMILARITY_TOP_K,
+        vector_store_query_mode: VectorStoreQueryMode = VectorStoreQueryMode.DEFAULT,
+        filters: Optional[MetadataFilters] = None,
+        alpha: Optional[float] = None,
+        node_ids: Optional[List[str]] = None,
+        doc_ids: Optional[List[str]] = None,
+        sparse_top_k: Optional[int] = None,
+        **kwargs: Any,
+    ) -> None:
+        """Initialize params."""
+        self._index = index
+        self._vector_store = self._index.vector_store
+        # separate image vector store for image retrieval
+        self._image_vector_store = self._index.image_vector_store
+        self._image_embed_model = self._index.image_embed_model
+        self._service_context = self._index.service_context
+        self._docstore = self._index.docstore
+
+        self._similarity_top_k = similarity_top_k
+        self._vector_store_query_mode = VectorStoreQueryMode(vector_store_query_mode)
+        self._alpha = alpha
+        self._node_ids = node_ids
+        self._doc_ids = doc_ids
+        self._filters = filters
+        self._sparse_top_k = sparse_top_k
+
+        self._kwargs: Dict[str, Any] = kwargs.get("vector_store_kwargs", {})
+
+    @property
+    def similarity_top_k(self) -> int:
+        """Return similarity top k."""
+        return self._similarity_top_k
+
+    @similarity_top_k.setter
+    def similarity_top_k(self, similarity_top_k: int) -> None:
+        """Set similarity top k."""
+        self._similarity_top_k = similarity_top_k
+
+    def _retrieve(
+        self,
+        query_bundle: QueryBundle,
+    ) -> List[NodeWithScore]:
+        res = self._text_retrieve(query_bundle)
+        res.extend(self._image_retrieve(query_bundle))
+        return res
+
+    def _text_retrieve(
+        self,
+        query_bundle: QueryBundle,
+    ) -> List[NodeWithScore]:
+        if self._vector_store.is_embedding_query:
+            if query_bundle.embedding is None and len(query_bundle.embedding_strs) > 0:
+                # get text embedding from Text embed model
+                query_bundle.embedding = (
+                    self._service_context.embed_model.get_agg_embedding_from_queries(
+                        query_bundle.embedding_strs
+                    )
+                )
+        return self._get_nodes_with_embeddings(query_bundle)
```

Comment:
is this the same as `super()._retrieve()`?

Reply:
it is, replaced :)